### DISCUSSION TITLE
How to maximise the Copilot's context awareness?

### DISCUSSION THREAD

**Discussion by nikitakunevich** [community/community#51323] [Copilot Conversations] [answered]
### Select Topic Area

Question

### Body

What context does github copilot use for generating suggestions?

Does it matter what files do i open in a VSCode or Intellij IDEA?
Or it is only the current file that is used as a context?

Knowing what factors are taken into account and what relative weights do they have would really help with using github copilot, as i'll know how to provide it with my project context better.

My findings:
- opening file for context significantly influence suggestions
- you don't have to open files for them to be used as a context
- regeneration of suggestions after opening/closing context files might significantly influence the results (maybe there is some caching of opened files involved)

**Comment by Akash1134** [comment, marked as answer]
Hi @nikitaKunevich ,

Copilot draws context from the code editor. It does not need access to your repositories. The code context sent to the model is commonly called a "prompt". We're constantly experimenting with new ways to create prompts that will tease even higher-quality suggestions out of the model.

As one proof point, Copilot had a suggestion acceptance rate of ~20% when we first launched the technical preview 18 months ago. Today, we're at 35% based largely on improvements to our prompts. We've simply gotten better at telling the model what kind of code the developer likely wants.

By way of example, one technique we use today is to look for relevant code in neighbouring files (i.e. tabs open in the code editor).
Imagine that you have two files open – one that defines a Class, and another with unit tests for that Class. By looking at both files, we have more context into the relevant code, which results in higher-quality suggestions.

In the future, you can imagine other prompt-crafting techniques that might produce even better quality suggestions. For example, what if we:
- Looked for relevant files in the project, but not open as active tabs
- Followed dependencies imported at the top of a file

Hope that helps!

**Reply by nikitakunevich** [reply]
Yep, it's very helpful.

So do you say that currently I can only expect the opened files in tabs or splits to be taken into account as a prompt?

**Reply by Akash1134** [reply]
Yes, for the current version of Copilot - this is true. We are continuously building the product to provide better quality suggestions so it won't be possible for us to share more context.

**Reply by einarpersson** [reply]
@Akash1134 

Is it still the case, that _**only open files**_ is provided as context (except the file itself)?

Or if I add a comment at the top `// This node-script will open and parse people.json` will the file `people.json` be included in the prompt? (assuming it is a sibling file)

I don't know whether there is a hesitancy to "share the secret sauce", but from a user/developer point of view it would be really useful to gain clarity and insight into how the current context is provided - even if it is shifting rapidly (i understand that you are actively developing and iterating upon this).

**Reply by mfp22** [reply]
I really really want to be able to have co-pilot refer to js doc examples for the functions I'm using.

**Reply by oxygenecore** [reply]
would be nice to have copilot pretrained (or better contextualized) based on github repos. This way there would be more incentive to write project documentation in github wikis or projects themselves, instead of using some external tools (like confluence).

**Reply by Ficik** [reply]
Does the "tabs open in the code editor" feature actually work? I heard it from multiple sources, but it never worked for me in vscode. Even in most simple version of having only 2 tabs open side-by-side, one with type interface and second where I'm defining variable of that type, it just guesses the interface.

Having context from other file that I'm actually looking at would be absolutely amazing.

**Reply by happyhardik** [reply]
I am happy and excited to hear that you are taking into consideration the open files for now. But the real fun is going to be when you start having the git repo into context. Please keep us updated when it's taken into your road map.

**Reply by umapriyadarsi-j-3266** [reply]
@nikitakunevich @Akash1134 how does it look in 2025 now?

My experience too has been that CoPilot is having more of an "Open files" only context. Even if I explicitly state in the prompt to refer another file it does not do it.

**Comment by Cornul11** [comment]
@einarpersson, I understand that while your statement is logically sound, publicly traded companies with shareholders would never reveal their algorithms or "trade secrets" to the public, even if we as developers and users would like to see them. This is due to obvious reasons, which is understandable.

**Comment by qiulang** [comment]
Hi @Akash1134,

It has been 1 year since you answered this question, I would like to know what the context scope copilot (v1.178.0) is using now.

In your answer you said "tabs open in the code editor", while the vscode document https://code.visualstudio.com/docs/copilot/overview 
 said 

> Copilot suggests code as you type. Copilot analyzes the context in the file you're editing and related files

What are the **related files** exactly ?

> Copilot has the context of your workspace and can give step-by-step guidance and code samples that are tailored to your project.
...
> Copilot suggests refactorings using the context of your codebase.
...
> Propose a fix for bugs in your code with the /fix command by suggesting code snippets and solutions based on the context of the error or issue. 

As it mentions the **workspace** can I assume it is not just the opened files ?

BTW, when codeium compare its features with copilot, https://codeium.com/ it says it supports "**Full repo context awareness**", but since my workspace is also my repo, so I am really confused with the context scope now.

<img width="1129" alt="codeium" src="https://github.com/community/community/assets/516243/6e29f9e8-6244-40f0-be26-b5fdfba576e5">

**Comment by Austin1serb** [comment]
While GitHub Copilot provides general code suggestions based on global coding patterns, I’m working on a tool that goes deeper by understanding your specific project. Instead of giving generic code suggestions, it scans your codebase, breaks down each file into relevant sections (Saves them as vector embed), and builds a structured understanding of your project. Then it fine-tunes the model with the data in your codebase. This way, it can provide suggestions and insights that are contextually aware of your project’s unique functions, components, and patterns—making the tool truly responsive to the specific code you’re working with. its going to be a lot of work. but I am working on Proof of concept right now
